### PR TITLE
feat(data-mapping): Implement Async Records Mapping Mechanism for MARC Migration (Process Chunks)

### PR DESCRIPTION
**Pull request by viacheslavkol**
### Purpose
Implement Async Records Mapping Mechanism for MARC Migration (Process Chunks)

### Approach
- Implement spring batch job with async step for chunks processing
- Read chunk entities from db
- Write step entities to db
- Read records for chunks
- Map records
- Update chunk/step entities in db
- Write data to local files
- Upload local files to minio after all batches processed

### Changes Checklist
- [x] **API Changes**: Document any API paths, methods, request or response bodies changed, added, or removed.
- [x] **Database Schema Changes**: Indicate any database schema changes and their impact. Confirm that migration scripts were created.
- [ ] **Interface Version Changes**: Indicate any changes to interface versions.
- [x] **Interface Dependencies**: Document added or removed dependencies.
- [x] **Permissions**: Document any changes to permissions.
- [x] **Logging**: Confirm that logging is appropriately handled.
- [x] **Unit Testing**: Confirm that changed classes were covered by unit tests.
- [x] **Integration Testing**: Confirm that changed logic was covered by integration tests.
- [x] **Manual Testing**: Confirm that changes were tested on local or dev environment.
- [x] **NEWS**: Confirm that the NEWS file is updated with relevant information about the changes made in this pull request.

### Related Issues
[MODMARCMIG-7](https://folio-org.atlassian.net/browse/MODMARCMIG-7)
[MODMARCMIG-12](https://folio-org.atlassian.net/browse/MODMARCMIG-12) - Task to support parallel migration operations processing for different tenants

### Testing
Testing was performed using docker-compose, two tenant setup, 4 threads for batch step execution, following env variables:
RECORDS_CHUNK_SIZE=10
CHUNK_FETCH_IDS_COUNT=20
CHUNK_PERSIST_COUNT=50
Data was pre-populated using testing sql scritps.
image
3 migration operation were submitted almost at the same time: 1 for 1'st tenant, then one for 2'nd tenant, 1 for 1'st tenant
All operations executed sequentially and finished successfully.
In a scope of operation - batches processed in parallel, confirmed by adding thread sleeps for even threads and thread name to logs.
image
image
image
Other tables for second tenant:
image
image
Minio for 2'nd tenant job:
image

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -91,6 +112,37 @@
...
+    <!-- Hibernate 6.4.1.Final has issue with saving or persisting JsonType fields and this was observed for authorities
+    when they are persisted hibernate increased version property twice (one for insert and one for update).
+    There is similar to this one issue in their issue tracker: https://hibernate.atlassian.net/browse/HHH-17660, we can
+    track its status and once it fixed we can try to upgrade version of hibernate which is enough to remove this dependency
+    and use the one from parent spring-boot -->
+    <dependency>
+      <groupId>org.hibernate.orm</groupId>
+      <artifactId>hibernate-core</artifactId>
+      <version>6.3.1.Final</version>
+    </dependency>
```
Comment (psmagin) [inline pom.xml:144]: Do we have JsonType fields? Or there is another cause of decreasing hibernate version?

File: src/main/java/org/folio/marc/migrations/config/ApplicationConfig.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.jdbc.core.BeanPropertyRowMapper;
+
+@Configuration
+public class ApplicationConfig {
+
+  @Bean
+  public BeanPropertyRowMapper<OperationChunk> operationChunkMapper() {
+    return new BeanPropertyRowMapper<>(OperationChunk.class);
```
Comment (psmagin) [inline src/main/java/org/folio/marc/migrations/config/ApplicationConfig.java:13]: Please provide doc why this bean is needed.

File: src/main/java/org/folio/marc/migrations/domain/entities/ChunkStep.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  @Enumerated(EnumType.STRING)
+  @JdbcTypeCode(SqlTypes.NAMED_ENUM)
+  @Column(nullable = false)
+  private StepStatus status;
+
+  @Column(name = "start_time_mapping")
+  private Timestamp stepStartTime;
+
+  @Column(name = "end_time_mapping")
+  private Timestamp stepEndTime;
```
Comment (psmagin) [inline src/main/java/org/folio/marc/migrations/domain/entities/ChunkStep.java:62]: Why do there start_time_mapping and end_time_mapping, can't it be step for saving (based on operationstep enum)

File: src/main/java/org/folio/marc/migrations/domain/entities/MarcRecord.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package org.folio.marc.migrations.domain.entities;
+
+import java.util.UUID;
+import org.folio.marc.migrations.domain.entities.types.RecordState;
+
+public record MarcRecord(UUID marcId, UUID authorityId, Object marc, RecordState state, Integer version) {
```
Comment (psmagin) [inline src/main/java/org/folio/marc/migrations/domain/entities/MarcRecord.java:6]: Should it be AuthorityMarcRecord or should authorityId be a recordId?

File: src/main/java/org/folio/marc/migrations/domain/entities/types/RecordState.java [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package org.folio.marc.migrations.domain.entities.types;
+
+public enum RecordState {
+  ACTUAL, DRAFT, OLD, DELETED
+}
```
Comment (psmagin) [inline src/main/java/org/folio/marc/migrations/domain/entities/types/RecordState.java:5]: Is it an enum for persistence? If not let's move it from "entities" package.
Can't we have only "ACTUAL" records based on view query?

File: src/main/java/org/folio/marc/migrations/domain/entities/MarcRecord.java (hunk 4 above)
Comment (psmagin) [reply src/main/java/org/folio/marc/migrations/domain/entities/MarcRecord.java:6]: Is it entitty for persistence? If not let's move it from "entities" package

**Review by psmagin — COMMENTED** [review]
Great job! 
It's hard to follow logic, but I think it's due not much experience with spring batch.
Will it be easy to extend this to instance/bib records?

File: pom.xml (hunk 1 above)
Comment (viacheslavkol) [reply pom.xml:144]: No, I guess it was added as part of attempt to fix some issue and wasn't removed after that. Will remove

**Comment by viacheslavkol** [comment]
> Great job! It's hard to follow logic, but I think it's due not much experience with spring batch. Will it be easy to extend this to instance/bib records?

I believe yes, classes are not authority dependent, except ChunkStepPreparationProcessor, which could be splitted to have smaller interchangable part for auth/bib, or additional job parameter could be used to define service to be used for retrieving appropriate entities

File: src/main/java/org/folio/marc/migrations/domain/entities/ChunkStep.java (hunk 3 above)
Comment (viacheslavkol) [reply src/main/java/org/folio/marc/migrations/domain/entities/ChunkStep.java:62]: That's a mistake, it's different in schema. It didn't break anything because not used with JPA. Will remove

File: src/main/java/org/folio/marc/migrations/domain/entities/MarcRecord.java (hunk 4 above)
Comment (viacheslavkol) [reply src/main/java/org/folio/marc/migrations/domain/entities/MarcRecord.java:6]: Changed `authorityId` to `recordId`.
Those are entities retrieved from database view. Or are you asking if we using them with JPA? If so - no

**Review by psmagin — APPROVED** [review]

**Review by TsaghikKhachatryan — APPROVED** [review]

**Review by viacheslavpoliakov — APPROVED** [review]
